src/parallel_parameter_search/walk_optimization.py

```diff
@@ -0,0 +1,508 @@
+# THIS HAS TO BE IMPORTED FIRST!!! I don't know why
+from bitbots_quintic_walk import PyWalk
+
+import math
+import time
+
+import dynamic_reconfigure.client
+import optuna
+import roslaunch
+import rospkg
+import rospy
+import tf
+from geometry_msgs.msg import Twist
+from nav_msgs.msg import Odometry
+
+from parallel_parameter_search.abstract_ros_optimization import AbstractRosOptimization
+from parallel_parameter_search.utils import set_param_to_file, load_yaml_to_param
+from parallel_parameter_search.simulators import PybulletSim, WebotsSim
+from sensor_msgs.msg import Imu, JointState
+
+
+class AbstractWalkOptimization(AbstractRosOptimization):
+
+    def __init__(self, namespace, robot_name, walk_as_node):
+        super().__init__(namespace)
+        self.rospack = rospkg.RosPack()
+        # set robot urdf and srdf
+        load_robot_param(self.namespace, self.rospack, robot_name)
+
+        # load walk params
+        load_yaml_to_param(self.namespace, 'bitbots_quintic_walk',
+                           '/config/walking_' + robot_name + '_optimization.yaml',
+                           self.rospack)
+
+        self.walk_as_node = walk_as_node
+        self.current_speed = None
+        self.last_time = 0
+        if self.walk_as_node:
+            self.walk_node = roslaunch.core.Node('bitbots_quintic_walk', 'WalkNode', 'walking',
+                                                 namespace=self.namespace)
+            self.walk_node.remap_args = [("walking_motor_goals", "DynamixelController/command"), ("/clock", "clock")]
+            self.launch.launch(self.walk_node)
+            self.dynconf_client = dynamic_reconfigure.client.Client(self.namespace + '/' + 'walking/engine', timeout=60)
+            self.cmd_vel_pub = rospy.Publisher(self.namespace + '/cmd_vel', Twist, queue_size=10)
+        else:
+            load_yaml_to_param("/robot_description_kinematics", robot_name + '_moveit_config',
+                               '/config/kinematics.yaml', self.rospack)
+            # create walk as python class to call it later
+            self.walk = PyWalk(self.namespace)
+
+        self.number_of_iterations = 10
+        self.time_limit = 10
+
+        # needs to be specified by subclasses
+        self.directions = None
+        self.reset_height_offset = None
+        self.reset_rpy_offset = [0, 0, 0]
+
+    def objective(self, trial):
+        raise NotImplementedError
+
+    def suggest_walk_params(self, trial):
+        raise NotImplementedError
```

Comment:
In the engine, much of the `suggest_walk_params` function happens in the abstract method to reduce code duplication. Would this be useful here, too?

Reply:
Don't know what you mean. we use a base method in stabilization and engine each. we can not put anything here, as the parameters are totally different